pallet-chainlink-feed/src/lib.rs

```diff
@@ -0,0 +1,1345 @@
+				let id: T::FeedId = FeedCounter::<T>::get();
+				ensure!(id < T::FeedLimit::get(), Error::<T>::FeedLimitReached);
+				let new_id = id.checked_add(&One::one()).ok_or(Error::<T>::Overflow)?;
+				FeedCounter::<T>::put(new_id);
+
+				let new_config = FeedConfig {
+					owner: owner.clone(),
+					pending_owner: None,
+					payment,
+					timeout,
+					submission_value_bounds,
+					submission_count_bounds,
+					decimals,
+					description,
+					restart_delay,
+					latest_round: Zero::zero(),
+					reporting_round: Zero::zero(),
+					first_valid_round: None,
+					oracle_count: Zero::zero(),
+				};
+				let mut feed = Feed::<T>::new(id, new_config); // synced on drop
+				let started_at = frame_system::Module::<T>::block_number();
+				let updated_at = Some(started_at);
+				// Store a dummy value for round 0 because we will not get useful data for
+				// it, but need some seed data that future rounds can carry over.
+				Rounds::<T>::insert(id, T::RoundId::zero(), Round {
+					started_at,
+					answer: Some(Zero::zero()),
+					updated_at,
+					answered_in_round: Some(Zero::zero())
+				});
+				feed.add_oracles(oracles)?;
+				// validate the rounds config
+				feed.update_future_rounds(payment, submission_count_bounds, restart_delay, timeout)?;
+				Self::deposit_event(RawEvent::FeedCreated(id, owner));
+				Ok(().into())
+			})
+		}
+
+		/// Initiate the transfer of the feed to `new_owner`.
+		#[weight = 100]
+		pub fn transfer_ownership(
+			origin,
+			feed_id: T::FeedId,
+			new_owner: T::AccountId,
+		) -> DispatchResultWithPostInfo {
+			let old_owner = ensure_signed(origin)?;
+			let mut feed = Self::feed_config(feed_id).ok_or(Error::<T>::FeedNotFound)?;
+			ensure!(feed.owner == old_owner, Error::<T>::NotFeedOwner);
+
+			feed.pending_owner = Some(new_owner.clone());
+			Feeds::<T>::insert(feed_id, feed);
+
+			Self::deposit_event(RawEvent::OwnerUpdateRequested(feed_id, old_owner, new_owner));
+
+			Ok(().into())
+		}
+
+		/// Accept the transfer of feed ownership.
+		#[weight = 100]
+		pub fn accept_ownership(
+			origin,
+			feed_id: T::FeedId,
+		) -> DispatchResultWithPostInfo {
+			let new_owner = ensure_signed(origin)?;
+			let mut feed = Self::feed_config(feed_id).ok_or(Error::<T>::FeedNotFound)?;
+
+			ensure!(feed.pending_owner.filter(|p| p == &new_owner).is_some(), Error::<T>::NotPendingOwner);
+
+			feed.pending_owner = None;
+			feed.owner = new_owner.clone();
+			Feeds::<T>::insert(feed_id, feed);
+
+			Self::deposit_event(RawEvent::OwnerUpdated(feed_id, new_owner));
+
+			Ok(().into())
+		}
+
+		/// Submit a new value to the given feed and round.
+		/// Limited to the oracles of a feed.
+		#[weight = 100]
+		pub fn submit(
+			origin,
+			feed_id: T::FeedId,
+			round_id: T::RoundId,
+			submission: T::Value,
+		) -> DispatchResultWithPostInfo {
+			let oracle = ensure_signed(origin)?;
+
+			with_transaction_result(|| -> DispatchResultWithPostInfo {
```

Comment:
why do we use transactional here?

Reply:
lots of reasons, actually :sweat_smile: 
but there are just lots of things storage mutations and lots of things that will error out, e.g. round initialization